builtin/providers/aws/resource_aws_eip_association.go

```diff
@@ -0,0 +1,157 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsEipAssociation() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsEipAssociationCreate,
+		Read:   resourceAwsEipAssociationRead,
+		Delete: resourceAwsEipAssociationDelete,
+
+		Schema: map[string]*schema.Schema{
+			"allocation_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"allow_reassociation": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"instance_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"network_interface_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"private_ip_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"public_ip": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsEipAssociationCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	request := &ec2.AssociateAddressInput{}
+
+	if v, ok := d.GetOk("allocation_id"); ok {
+		request.AllocationId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("allow_reassociation"); ok {
+		request.AllowReassociation = aws.Bool(v.(bool))
+	}
+	if v, ok := d.GetOk("instance_id"); ok {
+		request.InstanceId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("network_interface_id"); ok {
+		request.NetworkInterfaceId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("private_ip_address"); ok {
+		request.PrivateIpAddress = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("public_ip"); ok {
+		request.PublicIp = aws.String(v.(string))
+	}
+
+	log.Printf("[DEBUG] EIP association configuration: %#v", request)
+
+	resp, err := conn.AssociateAddress(request)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			return fmt.Errorf("[WARN] Error attaching EIP, message: \"%s\", code: \"%s\"",
+				awsErr.Message(), awsErr.Code())
+		}
+		return err
+	}
+
+	d.SetId(*resp.AssociationId)
+
+	return resourceAwsEipAssociationRead(d, meta)
+}
+
+func resourceAwsEipAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	request := &ec2.DescribeAddressesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String("association-id"),
```

Comment:
When an EIP is associated with something (ENI, Instance, Public IP etc), it get's an association-id. This was used as the ID. Therefore, the call to get the association uses the ID rather than the different scenarios it can be. An association_id will always be specific AFAICT